python/cudf/cudf/core/frame.py

```diff
@@ -438,128 +438,242 @@ def _normalize_columns_and_scalars_type(self, other):
                     )
                 )
 
-    def where(self, boolean_mask, other):
+    def where(self, cond, other=None, inplace=False):
         """
-        Replace values with other where the condition is False.
+        Replace values where the condition is False.
 
         Parameters
         ----------
-        cond : boolean
-            This can be Frame of booleans, where cond is True, keep the
-            original value. Where False, replace with corresponding value
-            from other. Callables are not supported.
-        other: list of scalars or a dataframe or a series,
+        cond : bool Series/DataFrame, array-like
+            Where cond is True, keep the original value.
+            Where False, replace with corresponding value from other.
+            Callables are not supported.
+        other: scalar, list of scalars, Series/DataFrame
             Entries where cond is False are replaced with
             corresponding value from other. Callables are not
-            supported.
+            supported. Default is None.
 
             DataFrame expects only Scalar or array like with scalars or
             dataframe with same dimention as self.
 
             Series expects only scalar or series like with same length
+        inplace : bool, default False
+            Whether to perform the operation in place on the data.
 
         Returns
         -------
-        result : DataFrame/Series
+        Same type as caller
 
         Examples:
         ---------
         >>> import cudf
         >>> df = cudf.DataFrame({"A":[1, 4, 5], "B":[3, 5, 8]})
-        >>> print (df.where(df % 2 == 0, [-1, -1]))
+        >>> df.where(df % 2 == 0, [-1, -1])
            A  B
         0 -1 -1
         1  4 -1
         2 -1  8
 
         >>> ser = cudf.Series([4, 3, 2, 1, 0])
-        >>> print(ser.where(ser > 2, 10))
+        >>> ser.where(ser > 2, 10)
         0     4
         1     3
         2    10
         3    10
         4    10
-        >>> print(ser.where(ser > 2))
-        0    4
-        1    3
-        2
-        3
-        4
+        dtype: int64
+        >>> ser.where(ser > 2)
+        0       4
+        1       3
+        2    null
+        3    null
+        4    null
+        dtype: int64
         """
 
         if isinstance(self, cudf.DataFrame):
-            if not isinstance(boolean_mask, cudf.DataFrame):
-                boolean_mask = self.from_pandas(pd.DataFrame(boolean_mask))
+            if isinstance(cond, cupy.core.ndarray):
```

Comment:
We shouldn't restrict this to just a cupy array. What if someone gave us a PyTorch tensor or a Numba NDArray?

We should check if this is an object that supports `__cuda_array_interface__`.

Reply:
This is now handled and also introduced tests for the same.